File: src/smartcasts/SmartCasts.java
package smartcasts;

import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import org.jetbrains.annotations.Nullable;

/**
 * @author Sergey Ryabov <colriot@gmail.com>
 *         20/06/2017
 */
public class SmartCasts {

  private static void printLength(@Nullable Object obj) {
    if (obj == null) {
      System.out.println("null");
    } else if (obj instanceof String) {
      System.out.println(((String) obj).length());
    } else if (obj instanceof Collection) {
      System.out.println(((Collection) obj).size());
    } else if (obj instanceof Date) {
      System.out.println(((Date) obj).getTime());
    } else {
      System.out.println(obj.hashCode());
    }
  }

  public static void main(String[] args) {
    printLength(null);
    printLength("string");
    printLength(Arrays.asList(1, 2, 3));
    printLength(new Date());
  }
}

File: src/smartcasts/SmartCasts.kt
package smartcasts

import java.util.*

/**
 * @author Sergey Ryabov <colriot@gmail.com>
 * 20/06/2017
 */
private fun printLength(obj: Any?) {
  if (obj == null) {
    println(null)
  } else if (obj is String) {
    println(obj.length)
  } else if (obj is Collection<*>) {
    println(obj.size)
  } else if (obj is Date) {
    println(obj.time)
  } else {
    println(obj.hashCode())
  }
}

fun main(args: Array<String>) {
  printLength(null)
  printLength("string")
  printLength(listOf(1, 2, 3))
  printLength(Date())
}